docs/design_proposals/private_pip_registries.md

```diff
@@ -0,0 +1,106 @@
+# Proposal Design: Support for Private `pip` registries
+
+## Objective
+
+The aim is to support private `pip` package registries in pixi.
+
+## Motivation
+
+To deliver and deploy internal software, organizations often host packages in private registries requiring authentication. Examples of such solutions are:
+
+* Google Cloud Artifact Registry
+* AWS CodeArtifact
+* Artifactory
+* packagecloud
+* ...etc.
+
+To use such repositories with pixi, authentication should be supported.
+
+## Design considerations
+
+1. **Compatibility**. Different package servers use different ways of deriving credentials. Some use static usernames and passwords, some use temporary tokens that need to be regenerated once in a while. Ideally, `pixi` itself should not be concerned about those details.
+2. **Ease of use**. In environments where authentication credentials can be derived automatically (e.g. GCP ADC), authentication should work "natively".
+3. **Configurable**. In `pixi.toml` file, developers should be able to choose what authentication mechanism to choose for a given channel.
+4. **Separation of concerns**. Authentication implementation for different platforms can be separate from `pixi` itself.
+
+## Proposed solution
+
+### Support for authentication helpers in `pixi`
+
+Authentication helper is a program that receives an URL to the package registry as an input and outputs a JSON object containing the credentials. Authentication helpers can utilize different methods for obtaining credentials, but it's expected that they are specific to the package host implementation.
+
+Authentication helper is a runnable binary available in `PATH`. The name of the binary must start with `pixi-auth-` and followed by its unique ID.
+
+A number of authentication helpers is shipped together with `pixi`. However, the interface between `pixi` and authentication helpers is documented, and it's expected that users may add a helper specific to their usecase.
+
+Authentication helpers are loosely inspired by `GIT_ASKPASS`, and authentication helpers in Docker and Flyte.
```

Comment:
Could you clarify this a little more, is your proposal to ship multiple binaries with together with pixi or are there commands that will be called. Or maybe pixi can install them when needed.

Reply:
Given that pixi is a package manager that can install packages, we could also ship them as conda dependencies of the project perhaps.

Reply:
I could see a future where you might have to update the helpers without updating the version of pixi. Wouldn't that make it easier if we support them as conda packages. The logic for "pre installing" is not really in yet. But we have something in the works for installing python before the pypi solve so it shouldn't be that hard to add.

Reply:
We could also implement this by having hard coded implementations and "fallback" commands to execute instead of having to ship multiple binaries. Having multiple binaries complicates the distribution of releases.

Reply:
I can see a future where we could have something like this:

```toml
[authentication-helpers]
gcloud = { spec = "conda-gcloud-auth-helper >=2.0.0" }
```

They could be part of the lock-file and installed before we do anything. That would require that helpers are available publicly of course.

Reply:
Im curious what that would look like! Do you mean something along the lines of `pixi auth-helper x`. If `x` is built-in it would use that but otherwise it tries to find the executable `pixi-auth-helper-x`?

I get the argument for now wanting to blow up the dependency footprint. I think careful consideration is required when deciding on which wants we want to bundle and which ones we don't. Installation of auth helpers should be fairly easy regardless because you could simply do `pixi global install pixi-super-cool-auth-helper`

Reply:
I have to think about this for a bit. 

We could easily encode the API version in a package that the authentication helper would need to depend on. When installing an environment for the auth helper we would simply include the requirement 'pixi-auth-helper-api=1.0'.

Reply:
To not get stuck on the distribution details before we have an working example. Lets continue with a separate binary that we might merge in or not when we decide what we do with this part. 

I think the implementation is not an overnight thing so I guess you could start on the implementation of supporting any (not just private) pypi index to get started.